### PR TITLE
Add methods for raw query params and update documentation

### LINKED ISSUE #14: URIInterface: withQuery requires an encoded string
It could be safer to rename it to withRawQuery (+ getRawQuery()), and make withQuery take `dict<string, ?string>` and require implementations to encode/escape, instead of application developers.

### PR DESCRIPTION
**Pull request by YuryBandarchuk16**
Fix #14

### CODE REVIEW

File: src/UriInterface.hh [hunk 1]
```diff
@@ -173,9 +173,21 @@ interface UriInterface {
...
+  /**
+   * Retrieve the query params of the URI.
+   *
+   * If no query params dict is present, this method MUST return an empty dict.
+   *
+   * The value returned MUST be the same as the value passed to withRawQuery()
+   *
+   * @return dict<string, ?string> The query params dict with unencoded (raw) values
+   */
+  public function getRawQuery(): dict<string, ?string>;
```
Comment (fredemmott) [inline src/UriInterface.hh]: I think this is the wrong way around: I would expect `getRawQuery()` to involve less processing than `getQuery()`

File: src/UriInterface.hh [hunk 2]
```diff
@@ -271,21 +283,26 @@ interface UriInterface {
   public function withPath(string $path): this;
 
   /**
-   * Return an instance with the specified query string.
+   * Return an instance with the specified query params dict.
```
Comment (fredemmott) [inline src/UriInterface.hh]: query parameters.

Specifying `dict` is redundant given it's in the signature

File: src/UriInterface.hh [hunk 3]
```diff
@@ -271,21 +283,26 @@ interface UriInterface {
...
+   * All query params which value is set to null MUST be omitted.
+   * The query params dict is assumed to be empty if there are no elements in
+   * it or all params have been omitted.
+   *
+   * An empty query params dict is equivalent to removing the query.
    *
-   * @param string $query The query string to use with the new instance.
-   * @return static A new instance with the specified query string.
-   * @throws \InvalidArgumentException for invalid query strings.
+   * @param dict<string, ?string> $query The query params.
```
Comment (fredemmott) [inline src/UriInterface.hh]: might as well just remove this, doesn't add anything over the signature

File: src/UriInterface.hh [hunk 4]
```diff
@@ -271,21 +283,26 @@ interface UriInterface {
...
+   * @return static A new instance with the specified query.
+   * @throws \InvalidArgumentException for invalid query params.
    */
-  public function withQuery(string $query): this;
+  public function withRawQuery(dict<string, ?string> $query): this;
```
Comment (fredemmott) [inline src/UriInterface.hh]: I don't think the `?string` type makes sense: what's the difference between:

- null
- not having the key set
- having it set to the empty string?

File: src/UriInterface.hh (hunk 1 above)
Comment (YuryBandarchuk16) [reply src/UriInterface.hh]: Could you tell me a bit more about what do you expect to see here?

In my opinion, the implementation of `getRawQuery()` just needs to return the dictionary that was passed to `withRawQuery()` so there is no processing needed.

File: src/UriInterface.hh (hunk 4 above)
Comment (YuryBandarchuk16) [reply src/UriInterface.hh]: Ok, seems reasonable. The I guess it worth mentioning that if the value equal to an empty string then the key will be omitted.

**Comment by fredemmott** [comment]
For example, ifthis is the current request URI, the true 'raw' query is the string - converting it to a dict is processing it.

`withRawQuery` should also be renamed to match.

**Comment by fredemmott** [comment]
I guess I'm meaning "raw" to mean "as in the HTTP and URI specifications", not "raw as set by hack code"

**Comment by fredemmott** [comment]
Yes, though not with those names.

Maybe:

```
function getRawQuery(): string;
function withRawQuery(string): this;
function getQuery(): dict<string, string>;
function withQuery(dict<string, string>);
```

.. so they'd do that with getQuery and withQuery, instead of *RawQuery

**Comment by YuryBandarchuk16** [comment]
Ok, I see. But we still need method to return percent-encoded query params string.

**Comment by fredemmott** [comment]
In that case, it would be `getRawQuery()`

File: src/UriInterface.hh [hunk 5]
```diff
@@ -172,8 +172,18 @@ interface UriInterface {
    *
    * @see https://tools.ietf.org/html/rfc3986#section-2
    * @see https://tools.ietf.org/html/rfc3986#section-3.4
+   * @return null|string The percent-encoded query string
    */
-  public function getQuery(): ?string;
+  public function getRawQuery(): ?string;
```
Comment (fredemmott) [inline src/UriInterface.hh]: > +   * If no query params or string is present, this method MUST return null


I'm not sure if this should be nullable actually - is there an observerable difference between requests to `"/foo"` and `"/foo?"` ?

**Pull request merged**